presto-hive/src/main/java/com/facebook/presto/hive/statistics/QuickStatsProvider.java

```diff
@@ -0,0 +1,308 @@
+        for (int counter = 0; counter < partitionIds.size(); counter++) {
+            String partitionId = partitionIds.get(counter);
+            partitionQuickStatCompletableFutures[counter] = supplyAsync(() -> getQuickStats(session, metastore, table, metastoreContext, partitionId), backgroundFetchExecutor);
+        }
+
+        try {
+            // Wait for all the partitions to get their quick stats
+            // If this query is reading a partition for which we do not already have cached quick stats,
+            // we will block the execution of the query until the stats are fetched for all such partitions,
+            // or we time out waiting for the fetch
+            // TODO : Make timeout value configurable
+            allOf(partitionQuickStatCompletableFutures).get(buildTimeoutMillis, MILLISECONDS);
+        }
+        catch (InterruptedException | ExecutionException e) {
+            log.error(e);
+            throw new RuntimeException(e);
+        }
+        catch (TimeoutException e) {
+            log.warn(e, "Timeout while building quick stats");
+            // TODO : Log a metric that we experienced a timeout
+        }
+
+        ImmutableMap.Builder<String, PartitionStatistics> result = ImmutableMap.builder();
+        for (int counter = 0; counter < partitionQuickStatCompletableFutures.length; counter++) {
+            String partitionId = partitionIds.get(counter);
+            CompletableFuture<PartitionStatistics> future = partitionQuickStatCompletableFutures[counter];
+            if (future.isDone() && !future.isCancelled() && !future.isCompletedExceptionally()) {
+                try {
+                    result.put(partitionId, future.get());
+                }
+                catch (InterruptedException | ExecutionException e) {
+                    // This should not happen because we checked that the future was completed successfully
+                    log.error(e, "Failed to get value for a quick stats future which was completed successfully");
+                    throw new RuntimeException(e);
+                }
+            }
+            else {
+                // If a future did not finish, or finished exceptionally, we do not add it to the results
+                // A new query for the same partition could trigger a successful quick stats fetch for this partition
+                result.put(partitionId, empty());
+            }
+        }
+
+        return result.build();
+    }
+
+    public PartitionStatistics getQuickStats(ConnectorSession session, SemiTransactionalHiveMetastore metastore, SchemaTableName table,
+            MetastoreContext metastoreContext, String partitionId)
+    {
+        if (!isQuickStatsEnabled(session)) {
+            return empty();
+        }
+        requestCount.incrementAndGet(); // New request was made to resolve quick stats for a partition
```

Comment:
I'm not returning `Cache#stats()` hit/miss counts because for this use case it is not painting the full picture. We also need to track the calls that could be answered by the inline call to the quick stats provider.